# Precipitation Data Preprocessing

This Jupyter Notebook is dedicated to the initial preprocessing of the historical precipitation data retrieved from NOAA FTP servers, transforming raw records into a clean and temporally consistent format suitable for subsequent hydrological analysis. The primary objective is to ensure the data's quality and integrity before its application in hydrological modeling.

The preprocessing steps undertaken in this notebook will primarily focus on **data cleaning and quality control**. This involves systematically addressing missing values,  and resolving inconsistencies within the 15-minute precipitation records to ensure the reliability of the input data. Identifying and correcting any detected outliers or erroneous readings will be performed on another notebook after the storm events have been selected. Furthermore, emphasis will be placed on ensuring **temporal consistency and alignment** across all records, verifying that the time series conform to a uniform resolution and are properly synchronized for accurate hydrological computations. The output of this notebook will be a refined dataset of quality-controlled and temporally consistent precipitation data, prepared for further processing such as storm event identification and ultimate use in driving the ModClark model.

In [113]:
# Import modules
import pandas as pd
from pathlib import Path
import warnings
import sys
import numpy as np
from datetime import datetime


# Supress warnings
warnings.filterwarnings("ignore")

In [41]:
# Import created functions
project_root = Path.cwd().parent.parent
if str(project_root) not in sys.path:
    sys.path.append(str(project_root))
from src.utils.data_utils import glimpse
from src.preprocessing.clean_preciptation_tabular_data import unpivot_preciptation_data

### Load and Inspect Precipitation Data

In [ ]:
# Build list of files
ppt_folder_path = project_root / 'data/bronze/tabular/precipitation'
ppt_file_list_path = list(ppt_folder_path.glob('*.csv'))
ppt_example_path = ppt_file_list_path[0]

In [20]:
# Inspect example file
ppt_example_df = pd.read_csv(ppt_example_path)
glimpse(ppt_example_df)

Rows: 10765
Columns: 491
           Null Count  Dtype    First Values
           ----------  -----    -------------
STATION    0           object   [USC00463669, USC00463669, USC00463669, USC00463669, USC00463669]
LATITUDE   0           float64  [38.4311, 38.4311, 38.4311, 38.4311, 38.4311]
LONGITUDE  0           float64  [-79.819, -79.819, -79.819, -79.819, -79.819]
ELEVATION  0           float64  [836.7, 836.7, 836.7, 836.7, 836.7]
DATE       0           object   [1992-06-02, 1992-06-03, 1992-06-04, 1992-06-05, 1992-06-06]
ELEMENT    0           object   [QPCP, QPCP, QPCP, QPCP, QPCP]
0000Val    0           int64    [-9999, 0, 0, 0, 0]
0000MF     0           object   [ , Z, Z, Z, Z]
0000QF     0           object   [M,  ,  ,  ,  ]
0000S1     0           object   [6, 6, 6, 6, 6]
0000S2     0           object   [ ,  ,  ,  ,  ]
0015Val    0           int64    [-9999, 0, 0, 0, 0]
0015MF     0           object   [ , Z, Z, Z, Z]
0015QF     0           object   [M,  ,  ,  ,  ]
0015S1     0  

The precipitation data is in wide-format. Each row represents a daily observation. The meaning of columns extracted from  documentation are given below:

- **HR00Val:** Is the value on the first hour of the day (i.e., the precipitation
           total during the time of day 00:00-01:00; missing = -9999).
           The units are _**hundredths of inch**_.
- **HR0MF**: is the measurement flag for the first hour of the day.
- **HR00QF**: is the quality flag for the first hour of the day.
- **HR00S1**: is the primary source flag for the first hour of the day.
- **HR00S2**: is the secondary source flag for the second hour of the day.

... and so on through the 24th hour of the day.

- **DlySum**: is the sum of all the non-missing, non-quality-control-flagged
           hourly values in the day (i.e., daily sum).
- **DlySumMF**: is the measurement flag placeholder that applies to the daily sum
           for consistency with the hourly value measurement flag.
           Currently this is always blank.
- **DlySumQF**: is the quality flag for the daily sum. It is blank if all 24 hourly
           values in the day were used to compute the sum,i
           and set to "P" (for Partial daily sum) if fewer than all 24 hourly
           values were used.
- **DlySumS1**: is the primary source flag for the daily sum for consistency with the
           hourly value primary source flag. Currently this is always blank.
- **DlySumS2**: is the secondary source flag for the daily sum for consistency with
           the hourly value secondary source flag. It is always set to "C" to
           indicate the daily sum is Computed from the hourly values preceding it in the same daily data record ... as opposed to a daily sum that might be reported by an observer who examines a daily accumulation amount in a rain gauge. 

In [58]:
print('Sample unique values for measurement flags:')
np.unique(np.concatenate([ppt_example_df['0000MF'].unique(), ppt_example_df['0015MF'].unique(), ppt_example_df['0030MF'].unique()]))

Sample unique values for measurement flags:


array([' ', '.', 'Z', 'a', 'g'], dtype=object)

The flag meanings, also extracted from the documentation, are as follow:

Blank = no measurement information applicable
- **g** = a carry-over measurement flag from the DSI-3240 dataset
                   which was used only on the very first hour of the month
                   if there was zero precipitation during that hour. The purpose
                   of this flag was mainly to indicate that the station was
                   functional and reporting during the month. Normally in
                   DSI-3240, **zero precipitation amounts were not included in the
                   data file in order to save space**. This HPD dataset does
                   include zero precipitation totals, both those assumed from
                   the DSI-3240 dataset and those determined from the digital
                   recordings of bucket level data. 
- **Z** = represents an "assumed" zero precipitation total. Usually
                   these are values from the DSI-3240 dataset. The rule in that
                   dataset was to "assume" a zero total for any hour where
                   nothing else was reported or indicated for that hour as long
                   as the very first hour of the month had a non-zero amount or
                   a zero amount with the "g" measurement flag. Zero amounts
                   were omitted from the DSI-3240 dataset in order to save disk
                   space. We are not concerned with that anymore.
- **a** = represents the beginning hour of an accumulation period.
                   Sometimes multi-hour accumulations were reported in DSI-3240,
                   so they were brought over for the period of record in which
                   we have DSI-3240 data when that was the best information
                   available. The data value for the beginning of an
                   accumulation period is set to missing (-9999).
- **.** = represents an hour during an accumulation period, between the
                   beginning and ending hours of the accumulation period.
                   The data value during the midst of an accumulation period is
                   set to missing (-9999). Accumulations across multiple hours
                   exist only from the legacy DSI-3240 data source. 
- **A** = designates the end of an accumulation period.
                   The accumulation total for the period is given for the
                   data value.
- **T** = trace of precipitation


In [59]:
print('Sample unique values for quality flags:')
np.unique(np.concatenate([ppt_example_df['0000QF'].unique(), ppt_example_df['0015QF'].unique(), ppt_example_df['0030QF'].unique()]))

Sample unique values for quality flags:


array([' ', 'D', 'M', 'R'], dtype=object)

Meanings for the quality flag, also extracted from documentation, are as follow:

- **Blank** = did not fail any quality assurance check
- **X**     = failed global extreme exceedence check
- **N**     = failed negative precipitation check
- **Y**     = failed state extreme exceedence check
        (performed on daily totals)
- **K**     = failed streak/frequent-value check
- **G**     = failed gap check
- **O**     = failed climatological outlier check
- **Z**     = flagged as a result of an official Datzilla investigation
- **A**     = The value is not an hourly precipitation total but rather
        an accumulation total for a period greater than an hour in
        duration and lasting through the end of this hour.
        (See measurement flags "a" and "." for the beginning time of
        the accumulation period and times during the accumulation
        period, respectively.) Accumulations across multiple hours
        exist only from the legacy DSI-3240 data source.
- **M**     = represents the associated value at this observation time is
        missing in the DSI-3240 dataset and no alternate data source
        is available. This is a carry-over indicator from DSI-3240 to
        allow the user to distinguish between missing and deleted
        data in that older system. (See the "D" quality flag.)
        However, the most consistent way to identify hours of missing
        data across the entire dataset is to test if the
        precipitation value is equal to the special missing value
        of -9999. 
- **D**     = represents the associated value at this time was deleted by
        the DSI-3240 processing system. Usually this was done
        manually by a trained meteorological technician who made the
        decision using ancillary information and experience.
- **Q**     = a carry-over quality flag the legacy DSI-3240 data source. 
        Prior to 1996: Indicates value failed an extreme value test
        (value will be present); data are to be used with caution.
        Extremes tests were:
        1) If the value was not an accumulated precipitation total,
            the value failed the one-hour statewide 100-year return
            period precipitation.
        2) If the value was an accumulated precipitation total,
            the value failed the 24-hour statewide extreme
            precipitation total. This flag was assigned during a 1997
            NCDC rehabilitation of the 1900-1995 DSI-3240 archive.
        Since January 1996: A single erroneous datum (value will be
        present). Lowest data resolution is hourly. This flag is
        rarely used in DSI-3240 since 1996.
- **q**     = a carry-over quality flag the legacy DSI-3240 data source. 
        Used since January 1996. An hourly value excludes one or
        more 15 minute periods. Lowest data resolution is 15 minutes.
- **R**     = a carry-over quality flag the legacy DSI-3240 data source.
        Used since January 1996. Indicates data values are suspect
        with regard to the times or period of occurrence. 


The following steps are taken to transform the raw tabular data, into a format suitable to initiate analysis:

1. Drops unnecessary columns: `['STATION', 'ELEMENT','LATITUDE', 'LONGITUDE', 'ELEVATION', 'DlySum', 'DlySumMF', 'DlySumQF', 'DlySumS1', 'DlySumS2']`.
2. Melts the DataFrame to long format based on the 'DATE' column.
3. Combines 'DATE' and 'Time_Attribute' for a new timestamp.
4. Extracts 'Attribute' from the 'Time_Attribute' column.
5. Pivots the data to separate out different measurements.
6. Renames the columns for clarity.
7. Replace -9999 (null values) by NA
8. Replace height values for failed quality cheks by NA
9. Filter out observation prior to year 2014 and after 2025-05-08, if any.
10. Fill in with NA values missing intervals

In [78]:
df = ppt_example_df.copy()
columns_to_trop = ['STATION', 'ELEMENT','LATITUDE', 'LONGITUDE', 'ELEVATION', 'DlySum', 'DlySumMF', 'DlySumQF', 'DlySumS1', 'DlySumS2']
df = df.drop(columns=columns_to_trop)

# Identifying columns to melt
id_vars = ['DATE']
value_vars = [col for col in df.columns if col not in id_vars]

# Melting to long format
df_long = df.melt(id_vars=id_vars,
                    value_vars=value_vars,
                    var_name='Time_Attribute',
                    value_name='Value')

# Add Time to DATE column and extract Attribute from 'Time_Attribute'
df_long['DATE'] = df_long['DATE'].astype(str) + ' ' + df_long['Time_Attribute'].str[:4]
df_long['Attribute'] = df_long['Time_Attribute'].str[4:]

df_long = df_long.drop(columns=['Time_Attribute'])

In [83]:
df_long.head()

DATE  Value Attribute
0  1992-06-02 0000  -9999       Val
1  1992-06-03 0000      0       Val
2  1992-06-04 0000      0       Val
3  1992-06-05 0000      0       Val
4  1992-06-06 0000      0       Val

In [84]:
df_long['Attribute'].unique()

array(['Val', 'MF', 'QF', 'S1', 'S2'], dtype=object)

In [85]:
df_long['Value'].unique()

array([-9999, 0, 10, 20, 30, 12, 1, 2, 4, 9, 3, 7, 5, 6, 13, 14, 8, ' ',
       'Z', 'g', '.', 'a', 'M', 'D', 'R', '6', 'H', 40, 21, 27, 19, 17,
       11, 16, 24, 26, 50, 15, 'A', 18, 49, 22, 60, 37, 'N', 29, 23, 34,
       25, 80, 70, 28, 116, 117, 35, 'Q', 114, 'q', 44, 68, 63, 115, 86,
       110, 38, 90, 100, 55, 41, 32, 57, 45, 58, 75, 33, 48, 42, 47, 43,
       51, 'G', 130, 36, 61, 31, 39], dtype=object)

In [86]:
# Pivoting to separate Value, Measurement Flag, Quality Flag, and Source Flags
df_final = df_long.pivot_table(
    index=['DATE'], 
    columns='Attribute', 
    values='Value', 
    aggfunc='first'
).reset_index()


In [88]:
df_final.head()

Attribute             DATE MF QF S1 S2    Val
0          1992-06-02 0000     M  6     -9999
1          1992-06-02 0015     M  6     -9999
2          1992-06-02 0030     M  6     -9999
3          1992-06-02 0045     M  6     -9999
4          1992-06-02 0100     M  6     -9999

In [89]:
df_final['MF'].unique()

array([' ', 'Z', 'g', 'a', '.', 'A'], dtype=object)

In [90]:
df_final['QF'].unique()

array(['M', ' ', 'A', 'D', 'R', 'G', 'Z', 'Q', 'q', 'N'], dtype=object)

In [91]:
# Rename  Columns
df_final.columns.name = None 
df_final = df_final.rename({
    'DATE':'date',
    'Val':'height',
    'MF':'measurement_flag',
    'QF':'quality_flag',
}, axis='columns')

# Drop columns
drop_columns = ['S1','S2']
df_final = df_final.drop(columns=drop_columns)

# Parse datetime column
df_final['date'] = pd.to_datetime(df_final['date'], format='%Y-%m-%d %H%M')

# Replace null values
df_final['height'] = df_final['height'].replace(-9999,pd.NA)

# Filter by date
cutoff_date = pd.to_datetime('20140101')
df_final = df_final[df_final['date']>=cutoff_date]

In [92]:
df_final.head()

date measurement_flag quality_flag height
655200 2014-01-01 00:00:00                                    0
655201 2014-01-01 00:15:00                                    0
655202 2014-01-01 00:30:00                                    0
655203 2014-01-01 00:45:00                                    0
655204 2014-01-01 01:00:00                                    0

In [93]:
df_final['measurement_flag'].unique()

array([' '], dtype=object)

In [94]:
df_final['quality_flag'].unique()

array([' ', 'N'], dtype=object)

In [95]:
glimpse(df_final)

Rows: 378240
Columns: 4
                  Null Count  Dtype           First Values
                  ----------  -----           -------------
date              0           datetime64[ns]  [2014-01-01T00:00:00.000000000, 2014-01-01T00:15:00.000000000, 2014-01-01T00:30:00.000000000, 2014-01-01T00:45:00.000000000, 2014-01-01T01:00:00.000000000]
measurement_flag  0           object          [ ,  ,  ,  ,  ]
quality_flag      0           object          [ ,  ,  ,  ,  ]
height            914         object          [0, 0, 0, 0, 0]


In [97]:
print(f'Sum of negative precipitation values: {sum(df_final['height']<0)}')

Sum of negative precipitation values: 0


In [100]:
df_final['height'] = df_final['height'].where(df_final['height'] >= 0, pd.NA)
df_final['height'].unique()

array([0, 1, 2, 5, 3, 4, 7, 6, 16, <NA>, 8, 12, 29, 10, 11, 9, 18, 15, 13,
       14, 19, 50, 20, 32, 75, 33, 27, 37, 47, 44, 17, 21, 63, 24, 49, 22,
       28, 31, 23, 36, 38, 26, 25, 39, 60, 30, 58, 57, 45, 35, 117, 86,
       68, 48, 42, 51, 43, 116, 114, 115, 61, 40, 55, 41], dtype=object)

In [102]:
df_final['height'].value_counts()

height
0      360522
1        8758
2        3312
3        1710
4         995
        ...  
115         1
61          1
40          1
55          1
41          1
Name: count, Length: 63, dtype: int64

In [104]:
df_final['measurement_flag'].value_counts()

measurement_flag
    378240
Name: count, dtype: int64

In [105]:
df_final['quality_flag'].value_counts()

quality_flag
     378190
N        50
Name: count, dtype: int64

In [106]:
df_final[df_final['quality_flag']=='N']

date measurement_flag quality_flag height
659472  2014-02-14 12:00:00                             N      0
659554  2014-03-06 08:30:00                             N      0
659555  2014-03-06 08:45:00                             N      0
773125  2017-08-03 09:15:00                             N      0
773126  2017-08-03 09:30:00                             N      0
773127  2017-08-03 09:45:00                             N      0
775901  2017-09-01 07:15:00                             N      0
775902  2017-09-01 07:30:00                             N      0
775903  2017-09-01 07:45:00                             N      0
777232  2017-09-15 04:00:00                             N      0
777234  2017-09-15 04:30:00                             N      0
777235  2017-09-15 04:45:00                             N      0
784644  2017-12-01 09:00:00                             N      0
784660  2017-12-01 13:00:00                             N      0
784661  2017-12-01 13:15:00                             N      0
784662  2017-12-01 13:30:00                             N      0
785234  2017-12-07 12:30:00                             N      0
785235  2017-12-07 12:45:00                             N      0
880856  2020-09-01 14:00:00                             N      0
880857  2020-09-01 14:15:00                             N      0
880859  2020-09-01 14:45:00                             N      0
880944  2020-09-02 12:00:00                             N      0
880945  2020-09-02 12:15:00                             N      0
880946  2020-09-02 12:30:00                             N      0
889696  2020-12-02 16:00:00                             N      0
901008  2021-04-01 12:00:00                             N      0
901009  2021-04-01 12:15:00                             N      0
901296  2021-05-04 12:00:00                             N      0
901298  2021-05-04 12:30:00                             N      0
901299  2021-05-04 12:45:00                             N      0
906950  2021-07-09 09:30:00                             N      0
906951  2021-07-09 09:45:00                             N      0
918197  2021-11-03 13:15:00                             N      0
918198  2021-11-03 13:30:00                             N      0
918199  2021-11-03 13:45:00                             N      0
918280  2021-11-04 10:00:00                             N      0
918282  2021-11-04 10:30:00                             N      0
918283  2021-11-04 10:45:00                             N      0
919644  2021-11-18 15:00:00                             N      0
919645  2021-11-18 15:15:00                             N      0
919646  2021-11-18 15:30:00                             N      0
935448  2022-05-02 06:00:00                             N      0
935449  2022-05-02 06:15:00                             N      0
935557  2022-06-04 09:15:00                             N      0
935558  2022-06-04 09:30:00                             N      0
935559  2022-06-04 09:45:00                             N      0
1025996 2025-01-01 11:00:00                             N      0
1025998 2025-01-01 11:30:00                             N      0
1025999 2025-01-01 11:45:00                             N      0
1033384 2025-04-01 10:00:00                             N      0

**Negative precipitation check**

The quality flag "N" indicates that negative precipitation check failed. Therefore, for those measurement, NA will be attributed for the height value.

In [107]:
# Replace failed to check negative precipitation with NA
df_final['quality_flag'] = df_final['quality_flag'].replace('N', pd.NA)

In [110]:
df_final.describe()

date
count                         378240
mean   2019-09-04 18:34:09.898476288
min              2014-01-01 00:00:00
25%              2016-12-01 23:56:15
50%              2019-08-17 23:52:30
75%              2022-07-08 23:48:45
max              2025-04-01 23:45:00

**Check if there are missing interval**

In [ ]:
def count_15min_intervals(date_min, date_max):
    if isinstance(date_min, str):
        date_min = datetime.strptime(date_min, '%Y%m%d %H:%M:%S')
    if isinstance(date_max, str):
        date_max = datetime.strptime(date_max, '%Y%m%d %H:%M:%S')
    delta = date_max - date_min
    total_minutes = delta.total_seconds() / 60
    return int(total_minutes // 15) + 1

In [121]:
missing_interval = count_15min_intervals(df_final['date'].min(), df_final['date'].max()) - df_final.shape[0]
print(f'Missing 15 minute intervals: {missing_interval}')

Missing 15 minute intervals: 16224


For this data example, there are over **16224** of missing interval. That's about half an year. Further  investigation will be perfomed to check the possiblity of zero precipitation data have been ommitted to save space. 

In [ ]:
# Define function to generate 15 minute sequence
def generate_15min_sequence(date_min, date_max):
    if isinstance(date_min, str):
        date_min = datetime.strptime(date_min, '%Y%m%d %H:%M:%S')
    if isinstance(date_max, str):
        date_max = datetime.strptime(date_max, '%Y%m%d %H:%M:%S')
    
    return pd.date_range(start=date_min, end=date_max, freq='15min')


In [144]:
time_seq = generate_15min_sequence(df_final['date'].min(), df_final['date'].max())
filled_seq = pd.DataFrame({'seq':time_seq}).merge(
    df_final,
    how='left',
    left_on='seq',
    right_on='date'
)

filled_seq['year'] = filled_seq['seq'].dt.year
missing_by_year = filled_seq.groupby('year')['date'].apply(lambda x: x.isna().sum())
print('Number of missing values by year:')
print(missing_by_year)

Number of missing values by year:
year
2014    7584
2015       0
2016     192
2017      96
2018       0
2019     288
2020     192
2021    3552
2022    3072
2023       0
2024       0
2025    1248
Name: date, dtype: int64


In [148]:
count_zeros_height_by_year = df_final.assign(year= df_final['date'].dt.year).groupby('year')['height'].apply(lambda x: sum(x==0))
print(f'Count of zero height values: {count_zeros_height_by_year}')

Count of zero height values: year
2014    26060
2015    33347
2016    33422
2017    33320
2018    32875
2019    33173
2020    33031
2021    30217
2022    30505
2023    33742
2024    33734
2025     7096
Name: height, dtype: int64


If the missing values, consistently was after an specific year, and zeros for data was missing for specific year, than it was possible zeros was ommitted from data. However, that's not what is shown above. There, the procedure is to fill those missing dates with NA for height values.

In [141]:
path = project_root / 'data/bronze/tabular/streamflow'
len(list(path.glob('*.parquet')))

586

In [118]:
df_final.shape

(378240, 4)